### PR TITLE
Add Property model

### PR DESCRIPTION
**Pull request by Gabrielpanga**
Add property model with relation

Issue: #38 / AYBP-5

### CODE REVIEW

File: src/models/property.model.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+export class Property extends BaseModel {
+  public id!: number;
+  public name!: string;
+  public minRent!: number;
+  public maxRent!: number;
+  public user!: User;
+  public readonly createdAt!: Date;
+  public readonly updatedAt!: Date;
+
+  public static async get(id: number): Promise<Property | null> {
```
Comment (Gabrielpanga) [inline src/models/property.model.ts]: What do you guys think about the includes:
- Current: Expose methods that will be used ignoring what includes.
- We use them directly on the controller? Each endpoint will have its own logic and probably will be repeated.
- Export a base FindOptions for that model, that could be reused on every search method.

File: tests/integration/controllers/user.controller.test.ts [hunk 2]
```diff
@@ -1,36 +1,110 @@
...
+        .post('/users')
+        .send(userFactory.build());
+      return expect(body.id).toBeTruthy();
+    });
+    it('returns 400 when parameters and not sent', async () => {
+      const { body } = await request(getApp())
+        .post('/users')
+        .send();
+      return expect(body.code).toBe(400);
+    });
```
Comment (kaiomagalhaes) [inline tests/integration/controllers/user.controller.test.ts]: please add an empty line here

File: tests/integration/controllers/user.controller.test.ts [hunk 3]
```diff
@@ -1,36 +1,116 @@
...
+
+    it(`when the user doens't exist it returns 404`, async () => {
+      const { body } = await request(getApp())
+        .patch(`/users/${invalidId}`)
+        .send();
+      return expect(body.code).toBe(404);
+    });
+
+    it('when invalid parameters it returns 400', async () => {
```
Comment (kaiomagalhaes) [inline tests/integration/controllers/user.controller.test.ts]: `with invalid params ...`

File: tests/integration/controllers/user.controller.test.ts [hunk 4]
```diff
@@ -1,36 +1,116 @@
...
+          userFactory.build({
+            name: userFactory.getFaker().random.alphaNumeric(129)
+          })
+        );
+      return expect(body.code).toBe(400);
+    });
+  });
+
+  describe('DELETE /users/:id', () => {
+    it('when the user exist it returns 1 on the response', async () => {
```
Comment (kaiomagalhaes) [inline tests/integration/controllers/user.controller.test.ts]: `the user existS`

File: tests/integration/models/base.model.test.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+describe('BaseModel', () => {
+  describe('#create', () => {
+    describe('with valid params', () => {
+      it('creates a model', async () => {
+        const baseModel = await BaseFactory.create();
+        expect(baseModel!.id).toBeTruthy();
+      });
+    });
+  });
```
Comment (kaiomagalhaes) [inline tests/integration/models/base.model.test.ts]: please add an empty line here

File: tests/integration/models/user.model.test.ts [hunk 6]
```diff
@@ -22,4 +22,47 @@ describe('User', () => {
...
+
+    describe('with invalid params', () => {
+      it('returns null', async () => {
+        const searchedUser = await User.findByName(`a user that doesn't exist`);
+        expect(searchedUser).toBeFalsy();
+      });
+    });
+  });
+
+  describe('#updateOne', () => {
```
Comment (kaiomagalhaes) [inline tests/integration/models/user.model.test.ts]: please update the tests here

File: tests/integration/models/user.model.test.ts [hunk 7]
```diff
@@ -22,4 +22,47 @@ describe('User', () => {
...
+    describe('with valid params', () => {
```
Comment (kaiomagalhaes) [inline tests/integration/models/user.model.test.ts]: 'with a valid name, it updates the user'

File: src/controllers/property.controller.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  @GET
+  @Response<IProperty>(200, 'Retrieve a property.')
+  @Response<NotFoundError>(404, 'property not found')
+  @Example<IProperty>(propertyExample)
+  async show(@PathParam('id') id: number): Promise<IProperty> {
+    const property = await Property.get(id);
+    if (property) {
+      return property;
+    }
+    throw new Errors.NotFoundError('property not found');
```
[the remark this answers is not in the document]
Comment (Gabrielpanga) [reply src/controllers/property.controller.ts]: We can, the only issue that i saw was that the HttpErrors are controllers error, related to Typescript rest library, and it didn't sound like an error we want to handle on the business layer.
Maybe we could create an error from our end and map it as 404 on the exceptionResolver. What do you think?

[the remark this answers is not in the document]
Comment (victor0402) [reply src/controllers/property.controller.ts]: Just noticed that the `Errors.NotFoundError` is from the ts library, right?

Yea, sounds good @Gabrielpanga

[the remark this answers is not in the document]
Comment (victor0402) [reply src/controllers/property.controller.ts]: @kaiomagalhaes thoughts?

**Pull request merged**